### PR TITLE
Vt parallise source detection

### PR DESCRIPTION
**Pull request by kenfus**
- Adds parallized source detection
- Adds functionality to split a karabo.image into multiple parts
- Adds mosaicking for source detection plots
- General improvements of the karabo image class

### CODE REVIEW

File: karabo/sourcedetection/result.py [hunk 1]
```diff
@@ -283,5 +378,250 @@ def get_peak_to_total_flux_variation_image(self) -> Image:
...
+        )
+        to_drop = self.__get_idx_of_overlapping_sources()
+        if len(to_drop) > 0:
+            print(f"Merged in total {len(to_drop)*2} sources into {len(to_drop)}")
+            # Create a boolean mask to keep sources not in to_drop
+            mask = np.ones(len(self._detected_sources), dtype=np.bool_)
+            mask[np.array(to_drop)] = False
+            self._detected_sources = self._detected_sources[mask]
+            self._detected_sources = cast(NDArray[np.float_], self._detected_sources)
+        return self._detected_sources
```
Comment (Lukas113) [inline karabo/sourcedetection/result.py:448]: To me this property seems suspicious. To me, it seems like you intended to write a lazy-property, becasue you've declared `self._detected_sources` in `__init__` as None. Normal lazy-properties are asking if it's None, then do some stuff, assign it to the protected instance-variable, and then return it. Or when already exists (is not None), return the instance variable instead. 

However, you assign `self._detected_sources` at multiple stages in the property function which doesn't make sense to me. Either you have a lazy property where you just ask if the instance-variable is None or not and do stuff dependent on that, OR you don't assign an instance-variable at all and do the processing in the property each time you call for it (in the second case you don't have to assign to `self._detected_sources` at all).

In addition, I personally dislike stdout-prints where you can't control if it's verbose or not. When something takes a while, ok. But then, `tqdm` would be a better choice in my opinion.

File: karabo/sourcedetection/result.py [hunk 2]
```diff
@@ -283,5 +378,250 @@ def get_peak_to_total_flux_variation_image(self) -> Image:
...
+    def __get_result_image(self, image_type: str) -> Image:
+        mi = ImageMosaicker()
+        images = [
+            getattr(result, f"get_{image_type}_image")()
+            for result in self.bdsf_detection
+        ]
+        return mi.process(images)[0]
+
+    def get_RMS_map_image(self) -> Image:
+        return self.__get_result_image("RMS_map")
```
Comment (Lukas113) [inline karabo/sourcedetection/result.py:459]: For each `get_*_image` function, how do you ensure that they work? In my opinion, either add then in a test (if it doesn't take too long), or define `image_type` as a Literal type. One of the mentioned methods should ensure, that these are valid functions.

File: karabo/sourcedetection/result.py [hunk 3]
```diff
@@ -283,5 +378,250 @@ def get_peak_to_total_flux_variation_image(self) -> Image:
...
+        )
+        # Check if some sources overlap
+        to_drop: List[int] = []
+        for i in range(len(combined_positions)):
+            for j in range(
+                i + 1, len(combined_positions)
+            ):  # Compare with subsequent sources to avoid repetition
+                dist = np.linalg.norm(combined_positions[i] - combined_positions[j])
+                if dist < self.min_pixel_distance_between_sources:
+                    print(f"Source {i} and {j} are being merged")
```
Comment (Lukas113) [inline karabo/sourcedetection/result.py:592]: I dislike verbose outputs for the mentioned reasons in a comment above. Especially inside a for-loop.

File: karabo/sourcedetection/result.py [hunk 4]
```diff
@@ -283,5 +378,250 @@ def get_peak_to_total_flux_variation_image(self) -> Image:
...
+        )
+        if len(to_drop) > 0:
+            print(f"Merged in total {len(to_drop)*2} sources into {len(to_drop)}")
+            # Create a boolean mask to keep sources not in to_drop
+            mask = np.ones(len(combined_positions), dtype=np.bool_)
+            mask[np.array(to_drop)] = False
+            combined_positions = combined_positions[mask]
+            combined_positions = cast(NDArray[np.float_], combined_positions)
+
+        return combined_positions
```
Comment (Lukas113) [inline karabo/sourcedetection/result.py:627]: To me this is looking suspiciously similar (almost identical) to the code in `detected_sources` property. I'm pretty sure that there is supposed to be a common basis where this kind of code exist just once.

File: karabo/test/data/restored.fits
Comment (Lukas113) [inline karabo/test/data/restored.fits:1]: 33.6MB adding to a github repo is a lot. Maybe putting it on cscs is a better approach?

If it should persist as part of the repository, it's path should be added in `karabo.test.conftest.TFiles`.

**Review by Lukas113 — CHANGES_REQUESTED** [review]
Overall nice changes.

Before merging, please address the requested changes.

File: karabo/imaging/image.py [hunk 5]
```diff
@@ -68,8 +105,9 @@ def write_to_file(
     ) -> None:
         """Write an `Image` to `path`  as .fits"""
         check_ending(path=path, ending=".fits")
-        if not os.path.exists(os.path.dirname(path)):
-            os.makedirs(os.path.dirname(path))
+        dir_name = os.path.dirname(path)
+        if dir_name != "" and not os.path.exists(dir_name):
```
[the remark this answers is not in the document]
Comment (kenfus) [reply karabo/imaging/image.py:109]: Sadly, when you do os.path.dir.name("test.fits"), it returns "". Thus that one is needed or then os.makedirs("") raises an error.

File: karabo/imaging/image.py [hunk 6]
```diff
@@ -136,6 +174,110 @@ def _update_header_after_resize(self) -> None:
...
+
+        Examples
+        --------
+        >>> # Assuming `self.data` is a 4D array with shape (C, Z, X, Y)
+        >>> # and `self.cutout` method is defined
+        >>> image = Image()
+        >>> cutouts = image.split_image(4, overlap=10)
+        >>> len(cutouts)
+        16  # because 4x4 grid
+        """
```
[the remark this answers is not in the document]
Comment (kenfus) [reply karabo/imaging/image.py:253]: That's OK, I did some test and negative values work well. I don't know why someone would want this, but it works.

[the remark this answers is not in the document]
Comment (Lukas113) [reply karabo/imaging/image.py:253]: As far as I can tell they shouldn't work well because you could then have a silent error (which are in my experience the worst). If you e.g take max(0, value), and it's value isn't supposed to be negative, then the code-execution of the following code could work, but not as intended.

File: karabo/imaging/image.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Lukas113) [reply karabo/imaging/image.py:109]: I don't understand what you're trying to say. Yes it returns "", but for that there is the abspath.

File: karabo/imaging/image.py [hunk 7]
```diff
@@ -25,27 +28,61 @@
...
+        self._fh_verbose = False
+        if path is not None:
+            self.path = path
+            self.data, self.header = fits.getdata(
+                str(self.path),
+                ext=0,
+                header=True,
+                **kwargs,
+            )
+        elif data is not None and header is not None:
```
[the remark this answers is not in the document]
Comment (kenfus) [reply karabo/imaging/image.py:54]: I tried to make  overload work:
```

    @overload
    def __init__(self, path: Union[str, FilePathType], **kwargs: Any) -> None:
        ...

    @overload
    def __init__(
        self, data: NDArray[np.float_], header: fits.header.Header, **kwargs: Any
    ) -> None:
        ...
```

However, it seems that this case is not doable with overload. I removed it and I am raising a clearer error. Maybe we can add a separate PR, if you really think overload here is good?

File: karabo/test/test_image.py [hunk 8]
```diff
@@ -135,14 +180,6 @@ def test_explore_sky():
 #     residual_image.save_to_file("result/residual.fits")
 #     sky.save_to_file("result/imaging_sky.txt")
 
-
-def test_power_spectrum(tobject: TFiles):
```
Comment (kenfus) [inline karabo/test/test_image.py:139]: Removed because useless test anyway.

File: karabo/sourcedetection/result.py (hunk 4 above)
Comment (kenfus) [reply karabo/sourcedetection/result.py:627]: Yep, done!

File: karabo/imaging/image.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Lukas113) [reply karabo/imaging/image.py:54]: This overload is not looking right, because you've to cover all possible combinations of the function signature with the overloads. We can have a look at it together if you like.

File: karabo/imaging/image.py [hunk 9]
```diff
@@ -351,20 +493,149 @@ def get_cellsize(self) -> np.float64:
...
+        hdu_in: Optional[Union[int, str]] = None,
+        hdu_weights: Optional[Union[int, str]] = None,
+        **kwargs: Any,
+    ):
+        """
+        Parameters
+        ----------
+        reproject_function : callable, optional
+            The function to use for the reprojection.
+        combine_function : {'mean', 'sum', 'median', 'first', 'last', 'min', 'max'}
```
[the remark this answers is not in the document]
Comment (Lukas113) [reply karabo/imaging/image.py]: You're right. However, I just looked at their source code. There I've just seen the mentioned combine-functions. But let me have a look at it again on Monday.

File: karabo/imaging/image.py [hunk 10]
```diff
@@ -351,20 +493,149 @@ def get_cellsize(self) -> np.float64:
...
+        **kwargs : dict, optional
+            Additional keyword arguments to be passed to the reprojection function.
+        """
+        self.reproject_function = reproject_function
+        self.combine_function = combine_function
+        self.match_background = match_background
+        self.background_reference = background_reference
+        self.hdu_in = hdu_in
+        self.hdu_weights = hdu_weights
+        self.kwargs = kwargs
```
[the remark this answers is not in the document]
Comment (Lukas113) [reply karabo/imaging/image.py:562]: Let us talk about this when we see each other at the office. I'm not convinced by that statement. To me, it doesn't make any sense why e.g. `kwargs` or `combine_function is consistent with multiple function-calls.

File: karabo/imaging/image.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Lukas113) [reply karabo/imaging/image.py]: Ah, I see what went wrong. It is as follows:

I've installed `v0.9.1` in my environment, probably constraint because we had to constrain `astropy <=5.2`, because we use `RASCIL 1.0.0` (they use private astropy-import). [reproject coadd.py <= v0.10.0](https://github.com/astropy/reproject/blob/v0.10.0/reproject/mosaicking/coadd.py) does just support the combine functions mentioned in my initial comment. However, from [reproject coadd.py >= v0.11.0](https://github.com/astropy/reproject/blob/v0.11.0/reproject/mosaicking/coadd.py) they enhanced their support for other combine functions, as described in the documentation you've referenced.

So, long story short. If we want to refer to a larger set of combine-functions, we would have to be sure, that we have `reproject >= v0.11.0` and therefore constrain it in our dev- and user dependencies. If it's resolvable, then you can constrain the version as mentioned. If not, you should just refer to `sum` and `median`, and update it once we have a more recent version of `reproject`.

[the remark this answers is not in the document]
Comment (Lukas113) [reply karabo/imaging/image.py]: An update on why `reproject` is constrained:

Because we have constrained `reproject <0.10` in our [Karabo-Feedstock rascil meta.yaml](https://github.com/i4Ds/Karabo-Feedstock/blob/main/rascil/meta.yaml), getting a higher version of `reproject` is not possible. I'm not aware what the reasons were to constrain the version that much (not sure how stable their api is, or how compatible it is with `RASCIL 1.0.0`). Thus, if you want to support a higher version, you have to build a dev-build of `RASCIL` which doesn't have that constraint, run the tests of Karabo, and if they're looking fine, make a new RASCIL-conda-wheel with a larger build-number. I can assist you in that process.

[the remark this answers is not in the document]
Comment (deiruch) [reply karabo/imaging/image.py]: IMHO I think it would be sensible to put version constraints on Karabo itself, but not on the libraries in the feedstock.

[the remark this answers is not in the document]
Comment (deiruch) [reply karabo/imaging/image.py]: Yes, exactly like this: Loose versioning for Feedstock, locked-down versions for Karabo.

**Comment by kenfus** [comment]
> Overall nice changes.
> 
> Before merging, please address the requested changes.

Done! Please have another look.

File: karabo/imaging/image.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (kenfus) [reply karabo/imaging/image.py]: We will do this in a separate PR

File: karabo/sourcedetection/result.py [hunk 11]
```diff
@@ -283,5 +364,357 @@ def get_peak_to_total_flux_variation_image(self) -> Image:
...
+                    beam=beam,
+                    quiet=quiet,
+                    **kwargs,
+                )
+            )
+        if DaskHandler.dask_client is not None:
+            results = compute(*results, scheduler="distributed")
+        print(results[0])
+
+        return PyBDSFSourceDetectionResultList(results)
```
Comment (Lukas113) [inline karabo/sourcedetection/result.py:468]: Actually, type-safety is not ensured here. The reason why `mypy` sadly isn't complaining here is because `compute` is a Union which `Any` which isn't checked by design.

It is also a bad habit to assign a different type to the same variable `results`, which is happening at `results = compute(*results, scheduler="distributed")`.

Without the dask-compute, `mypy` should have complained about that you don't have `List[PyBDSFSourceDetectionResult] | None`. The reason is, because you didn't declare the list to be of type `List[PyBDSFSourceDetectionResult]` at definition. This would have ensured that what you did in the for-loop was correct. It's actually not safe what's done in the loop because `detect_sources_in_image` can return a `None` type, which would lead to a `List[PyBDSFSourceDetectionResult | None]` type which is not `List[PyBDSFSourceDetectionResult] | None` as declared in the `PyBDSFSourceDetectionResultList` constructor.

Long story short: Please rewrite this function  that it ensures type-safety to prevent unchecked run-time bugs in the future.

File: karabo/sourcedetection/result.py [hunk 12]
```diff
@@ -283,5 +364,357 @@ def get_peak_to_total_flux_variation_image(self) -> Image:
...
+            of the sources in the mosaic image.
+        """
+        if self.bdsf_detection is None:
+            raise ValueError(
+                "No PyBDSF detection results found. Did you run "
+                "detect_sources_in_images()?"
+            )
+        # Get headers
+        headers = [
+            result.get_source_image().header  # type: ignore
```
Comment (Lukas113) [inline karabo/sourcedetection/result.py:647]: This isn't save code because `get_source_image` can be `None` which has no `header` field. I'm not entirely sure how to exactly handle despite just cheaply raise a ValueError in a for-loop. However, I'm not sure if that's the way this function is supposed to handle if there is no source-image in a list of `Image`(s).

File: karabo/sourcedetection/result.py [hunk 13]
```diff
@@ -283,5 +364,357 @@ def get_peak_to_total_flux_variation_image(self) -> Image:
...
+        if self.bdsf_detection is None:
+            raise ValueError(
+                "No PyBDSF detection results found. Did you run "
+                "detect_sources_in_images()?"
+            )
+        # Get XY pixel position for each result
+        xy_poss = [
+            result.get_pixel_position_of_sources() for result in self.bdsf_detection
+        ]
+        assert all([result.has_source_image() for result in self.bdsf_detection])
```
Comment (Lukas113) [inline karabo/sourcedetection/result.py:694]: I think raising an Error instead of doing an assert is a nicer feedback as an error for a user.

**Review by Lukas113 — CHANGES_REQUESTED** [review]

File: karabo/sourcedetection/result.py [hunk 14]
```diff
@@ -83,6 +146,7 @@ def detect_sources_in_image(
                 format="csv",
                 **kwargs,
             )
+            return cls(detection)  # type: ignore
```
[the remark this answers is not in the document]
Comment (kenfus) [reply karabo/sourcedetection/result.py:160]: Was already there before. It raises this error otherwise:

`karabo/sourcedetection/result.py:160: error: Missing positional argument "source_image" in call to "SourceDetectionResult"  [call-arg]`

Like I said, source detection should be restructured anyway because it still works but mypy raises a weird error because of the usage of cls and some weird inheritance.

File: karabo/sourcedetection/result.py [hunk 15]
```diff
@@ -49,23 +70,65 @@ def __init__(
...
+        If the 'beam' parameter is not provided, the method will attempt to extract the
+        beam parameters from the image metadata. A warning is raised if beam parameters
+        are not found.
+
+        The PyBDSF process_image function is called for source detection, which is
+        particularly designed for radio astronomical images. For details on this
+        function, refer to the PyBDSF documentation.
         """
-        if beam is None:
+        if beam is None and not isinstance(image, List):
```
[the remark this answers is not in the document]
Comment (kenfus) [reply karabo/sourcedetection/result.py]: Done!

File: karabo/sourcedetection/result.py (hunk 12 above)
Comment (kenfus) [reply karabo/sourcedetection/result.py:650]: The design of sourcedetection is terrible because functions don't work if `detect_sources_in_images` was not called. Like I said #528

**Comment by kenfus** [comment]
@Lukas113 Please have another look. I would like to later restructure source detection completely, see #528

File: karabo/imaging/image.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Lukas113) [reply karabo/imaging/image.py]: Currently not correctly specified yet. Removing the `.*` should do the job.
```
- reproject               >=0.9, <=10.0
```

File: karabo/sourcedetection/result.py [hunk 16]
```diff
@@ -283,5 +364,357 @@ def get_peak_to_total_flux_variation_image(self) -> Image:
...
+        self.verbose = False
+        self.bdsf_detection = bdsf_detection
+
+    @staticmethod
+    def detect_sources_in_images(
+        images: List[Image],
+        beams: Optional[List[BeamType]] = None,
+        quiet: bool = True,
+        **kwargs: Any,
+    ) -> Optional[PyBDSFSourceDetectionResultList]:
```
[the remark this answers is not in the document]
Comment (kenfus) [reply karabo/sourcedetection/result.py:433]: No, sometimes Pybdsf returns None

File: karabo/sourcedetection/evaluation.py [hunk 17]
```diff
@@ -24,7 +27,7 @@ def __init__(
         ground_truth: NDArray[np.float_],
         assignments: NDArray[np.float_],
         sky_idxs: NDArray[np.int_],
-        source_detection: SourceDetectionResult,
+        source_detection: Union[SourceDetectionResult, PyBDSFSourceDetectionResultList],
```
[the remark this answers is not in the document]
Comment (kenfus) [reply karabo/sourcedetection/evaluation.py:30]: I am unsure how it will look after restructuring, thus I would like to leave it like this for now.

File: karabo/sourcedetection/result.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (Lukas113) [reply karabo/sourcedetection/result.py:160]: Ah, you're right. The issue issue is actually because the signature of the super- and subclass don't match which is actually a bug. We just have discovered it because we have typed the TypeVar of the function correctly. However, as you say, this is out of scope for your PR since it's not from your PR.

However, this needs to be fixed ASAP.

File: karabo/sourcedetection/evaluation.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (Lukas113) [reply karabo/sourcedetection/evaluation.py:30]: I strongly suggest to do so because it's done pretty quickly end we don't have to start making exceptions. I'll provide you with a minimal example:

```python
from abc import ABC, abstractmethod

import numpy as np
from numpy.typing import NDArray
from typing import Optional

from karabo.imaging.image import Image


class ISourceDetectionResult(ABC):
    @property
    @abstractmethod
    def detected_sources(self) -> NDArray[np.float_]:
        ...

    @abstractmethod
    def has_source_image(self) -> bool:
        ...

    @abstractmethod
    def get_source_image(self) -> Optional[Image]:
        ...

```

then add property where they're not explicitly defined. In this case, they're not defined in `SourceDetectionResult`. Just do the following:
```python
# in constructor
self._detected_sources = detected_sources

# add properties in class
@property
def detected_sources(self) -> NDArray[np.float_]:
    return self._detected_sources
```

then let the according classes inherit from the abstract base-class, 
```python
SourceDetectionResult(ISourceDetectionResult, KaraboResource):
PyBDSFSourceDetectionResultList(ISourceDetectionResult):
```

And then, just define the attribute in `SourceDetectionEvaluation` as follows:
```python
from karabo.sourcedetection.result import ISourceDetectionResult

# change type in __init__
source_detection: ISourceDetectionResult
```

I think I don't have forgotten anything.

File: karabo/sourcedetection/result.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (Lukas113) [reply karabo/sourcedetection/result.py:433]: Ah, I see. Not it's done properly with your changes.

File: karabo/sourcedetection/result.py (hunk 12 above)
Comment (Lukas113) [reply karabo/sourcedetection/result.py:650]: Ok. But then please code it safely within a for-loop and just raise an ValueError. Anything else is not safe and IMO we shouldn't start just adding `type: ignore` when the issue is easy resolvable.

File: karabo/sourcedetection/evaluation.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (Lukas113) [reply karabo/sourcedetection/evaluation.py:30]: I think we should get rid of the not correctly specified interface `KaraboResource` at some point (not in this PR).

[the remark this answers is not in the document]
Comment (kenfus) [reply karabo/sourcedetection/evaluation.py:30]: Seems like some work for something which will be restructured anyway. I can take a look at this next week, but I rather would restructure source detection completely in that time.

File: karabo/sourcedetection/result.py (hunk 12 above)
Comment (kenfus) [reply karabo/sourcedetection/result.py]: Done

File: karabo/sourcedetection/evaluation.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (Lukas113) [reply karabo/sourcedetection/evaluation.py]: fixed with 5be5377816a230d9998ef7b385b29305750c681c

**Review by Lukas113 — APPROVED** [review]
LGTM now
